### PR TITLE
Pdf & qr code stuffs

### LINKED ISSUE #91: Thoughts on Roles/Permissions modules
In the process of looking at #87, I took another look at the Roles/Permissions modules and had some thoughts.

* [x] Let's make your suggested change of suffixing `Enum` on the end of enumerations
* [x] Inputs and outpus are all in terms of _database IDs_ (implementation detail) instead of _values_ (conceptually more natural). For example it would be good if I could call `Roles.fromUser` and get a value (`'VISITOR'` or `RolesEnum.VISITOR`) instead of a number that describes that value.
* [x] The `ExistsQueryResponse` type should really just be `boolean`
* [x] The `QueryResponse` type doesn't need to be a union type, because `T` in `Promise<T>` is only the type of the resolve value.

Another thing I noticed is that in the Auth strategy definition, the roles/permissions interfaces are being initialised _on each request_. I didn't catch this on review (my bad), but what I intended was for the interfaces each to be initialised once on startup, and then stored to be referenced later (much like the `knex` client is). That was kind of the whole (poorly expressed) point of that pattern; you're binding a variable to a group of functions so that you don't need to pass it in each time.

In general I still kind of like that pattern, however, looking at the code, I can see it actually reduces the possibility for re-use. Example: the `move` method is basically a combination of `userHas`, `remove` and `add`, but it's not possible to call the latter two because they need to be done as a transaction. Passing in the knex client allows us to pass in `Knex.Transaction` instances as well as `Knex` or `Knex.QueryBuilder` instances.

What do you think?

### PR DESCRIPTION
**Pull request by eliascodes**
Fixes #91 

Work done in the process of writing the visitor registration route.

### Changes
* Add constraints on user roles (one role per user, one org per user)
* Add QR code service
* Add PDF generator service

### CODE REVIEW

File: src/utils/index.ts [hunk 1]
```diff
@@ -24,3 +26,25 @@ export const renameKeys: RenameKeys<Dictionary<any>, Dictionary<string>> =
     Object.keys(o)
       .reduce((acc, k) => assoc(map[k] || k, o[k], acc), {})
 );
+
+export const pipeToPromise = (stream: stream.Readable): Promise<Buffer[]> =>
```
Comment (astroash) [inline src/utils/index.ts]: curious of why this this is called `pipeToPromise`?

File: src/services/qrcode/__tests__/index.test.unit.ts [hunk 2]
```diff
@@ -0,0 +1,18 @@
+import { createHmac } from 'crypto';
+import * as QRCode from '..';
+import { getConfig } from '../../../../config';
+
+
+describe('QRCode Service', () => {
+  const config = getConfig(process.env.NODE_ENV);
+
+  describe('create', () => {
```
Comment (astroash) [inline src/services/qrcode/__tests__/index.test.unit.ts]: Do we need a fail case here?

File: src/utils/index.ts (hunk 1 above)
Comment (eliascodes) [reply src/utils/index.ts]: Streams have a method `pipe` which allows you to write the output of a `Readable` stream directly into a `Writable` stream. But I want to `pipe` the contents of a `Readable` stream into memory, and simply be notified (via promise resolution) when that's done.

File: src/services/qrcode/__tests__/index.test.unit.ts (hunk 2 above)
Comment (eliascodes) [reply src/services/qrcode/__tests__/index.test.unit.ts]: I don't think I can force a failure here (without annoying the compiler). I'm just calling 3rd party functions.

File: src/services/pdf/__tests__/index.test.unit.ts [hunk 3]
```diff
@@ -0,0 +1,17 @@
+import * as path from 'path';
+import { getConfig } from '../../../../config';
+import { PdfTemplateEnum, fromTemplate } from '..';
+
+
+describe('PDF Service', () => {
+  const { root } = getConfig(process.env.NODE_ENV);
+
+  describe('fromTemplate', () => {
+    test('Visitor QR Code Template', async () => {
```
[the remark this answers is not in the document]
Comment (eliascodes) [reply src/services/pdf/__tests__/index.test.unit.ts]: Yes, probably. The only fail case I can reasonably create is by passing the wrong dictionary shape into the function.

File: src/utils/index.ts (hunk 1 above)
Comment (astroash) [reply src/utils/index.ts]: Can this be renamed as `streamPipeToPromise` for clarity?

Comment (eliascodes) [reply src/utils/index.ts]: `pipe` is the verb here, so it should be first really. Can do `pipeStreamToPromise` if you like.

**Pull request merged**